packages/api/src/controllers/usage.js

```diff
@@ -0,0 +1,89 @@
+import Router from "express/lib/router";
+import { db } from "../store";
+
+const app = Router();
+
+app.get("/", async (req, res) => {
+  if (!req.headers.authorization) {
+    res.status(403);
+    return res.json({ errors: ["unauthorized"] });
+  }
+
+  let token = req.headers.authorization.split(" ")[1]; // Bearer <token>
+
+  if (process.env.LP_API_ADMIN_TOKEN != token) {
+    res.status(403);
+    return res.json({ errors: ["unauthorized"] });
+  }
+
+  let { fromTime, toTime } = req.query;
+
+  // if time range isn't specified return all usage
+  if (!fromTime && !toTime) {
+    fromTime = +new Date(2020, 0); // start at beginning
+    toTime = +new Date();
+  }
+
+  const cachedUsageHistory = await db.stream.cachedUsageHistory(
+    fromTime,
+    toTime,
+    {
+      useReplica: true,
+    }
+  );
+
+  res.status(200);
+  res.json(cachedUsageHistory);
+});
+
+app.post("/update", async (req, res) => {
+  if (!req.headers.authorization) {
+    res.status(403);
+    return res.json({ errors: ["unauthorized"] });
+  }
+
+  let token = req.headers.authorization.split(" ")[1]; // Bearer <token>
+
+  if (process.env.LP_API_ADMIN_TOKEN != token) {
+    res.status(403);
+    return res.json({ errors: ["unauthorized"] });
+  }
+
+  let fromTime = +new Date(2020, 0); // start at beginning
+  let toTime = +new Date();
+
+  let cachedUsageHistory = await db.stream.cachedUsageHistory(
+    fromTime,
+    toTime,
+    {
+      useReplica: true,
+    }
+  );
+
+  // get last updated date from cache
+  if (cachedUsageHistory.length) {
+    fromTime = cachedUsageHistory[cachedUsageHistory.length - 1].date;
+  }
+
+  // get all usage up until now
+  toTime = new Date().getTime();
+
+  let usageHistory = await db.stream.usageHistory(fromTime, toTime, {
+    useReplica: true,
+  });
+
+  // store each day of usage
+  for (const row of usageHistory) {
```

Comment:
Do we need to loop through all rows?

Seems like the only case where a row might already exist in the cache is for the first row returned since there could be overlap between the last day in the cache and the first day in the new usage history rows?

Reply:
I think so. If the cache hasn't been updated in a few days for some reason (ie the cron job is down) then we'll want to create rows for each day beginning at the last updated date.

Reply:
Ah I think I misread this before. We loop through all the rows in usageHistory because we need to create corresponding rows in the usage table. And if the fromTime matches the date of an existing row in the usage table then that is the one scenario where we replace the row with the new row instead of creating a row. Got it.